third_party/libusb/webport/src/usb_transfer_destination.cc

```diff
@@ -26,34 +26,123 @@
+      break;
+    case chrome_usb::ControlTransferInfoRecipient::kEndpoint:
+      request_type |= LIBUSB_RECIPIENT_ENDPOINT;
+      break;
+    case chrome_usb::ControlTransferInfoRecipient::kOther:
+      request_type |= LIBUSB_RECIPIENT_OTHER;
+      break;
+  }
+  return request_type;
+}
+
+}  // namespace
+
 UsbTransferDestination::UsbTransferDestination() = default;
 
 UsbTransferDestination::~UsbTransferDestination() = default;
 
+// static
+UsbTransferDestination UsbTransferDestination::CreateForControlTransfer(
+    int64_t js_device_handle,
+    uint8_t request_type,
+    uint8_t request,
+    uint16_t value,
+    uint16_t index) {
+  return UsbTransferDestination(js_device_handle,
+                                /*endpoint_address=*/{}, request_type, request,
+                                value, index);
+}
+
+// static
+UsbTransferDestination UsbTransferDestination::CreateForGenericTransfer(
+    int64_t js_device_handle,
+    uint8_t endpoint_address) {
+  return UsbTransferDestination(
+      js_device_handle, endpoint_address,
+      /*control_transfer_request_type=*/{}, /*control_transfer_request=*/{},
+      /*control_transfer_value=*/{}, /*control_transfer_index=*/{});
+}
+
 // static
 UsbTransferDestination
 UsbTransferDestination::CreateFromChromeUsbControlTransfer(
     const chrome_usb::ConnectionHandle& connection_handle,
     const chrome_usb::ControlTransferInfo& transfer_info) {
   return UsbTransferDestination(
-      connection_handle, transfer_info.direction, {}, transfer_info.recipient,
-      transfer_info.request_type, transfer_info.request, transfer_info.value,
-      transfer_info.index);
+      connection_handle.handle,
+      /*endpoint_address=*/{}, GetLibusbRequestType(transfer_info),
+      transfer_info.request, transfer_info.value, transfer_info.index);
 }
 
 // static
 UsbTransferDestination
 UsbTransferDestination::CreateFromChromeUsbGenericTransfer(
     const chrome_usb::ConnectionHandle& connection_handle,
     const chrome_usb::GenericTransferInfo& transfer_info) {
-  return UsbTransferDestination(connection_handle, transfer_info.direction,
-                                transfer_info.endpoint, {}, {}, {}, {}, {});
+  return UsbTransferDestination(
+      connection_handle.handle, transfer_info.endpoint,
+      /*control_transfer_request_type=*/{}, /*control_transfer_request=*/{},
+      /*control_transfer_value=*/{}, /*control_transfer_index=*/{});
 }
 
 bool UsbTransferDestination::IsInputDirection() const {
-  return direction_ == chrome_usb::Direction::kIn;
+  // It's only correct to check the flag presence by doing bitwise and against a
+  // non-zero constant.
+  static_assert(LIBUSB_ENDPOINT_IN != 0,
+                "Bad mask: LIBUSB_ENDPOINT_IN is zero");
+
+  if (control_transfer_request_type_) {
+    // For control transfers, the direction is encoded in the request type.
+    return (*control_transfer_request_type_ & LIBUSB_ENDPOINT_IN) != 0;
```

Comment:
As discussed, better to use (*control_transfer_request_type_ & LIBUSB_ENDPOINT_DIR_MASK) == LIBUSB_ENDPOINT_IN here